# 1.0 `About Dataset`
Context
While many public datasets (on Kaggle and the like) provide Apple App Store data, there are not many counterpart datasets available for Google Play Store apps anywhere on the web. On digging deeper, I found out that iTunes App Store page deploys a nicely indexed appendix-like structure to allow for simple and easy web scraping. On the other hand, Google Play Store uses sophisticated modern-day techniques (like dynamic page load) using JQuery making scraping more challenging.

### 1.1 Content
Each app (row) has values for catergory, rating, size, and more.

### 1.2 Acknowledgements
This information is scraped from the Google Play Store. This app information would not be available without it.

### 1.3 Inspiration
The Play Store apps data has enormous potential to drive app-making businesses to success. Actionable insights can be drawn for developers to work on and capture the Android market!

### 1.4 Dataset link:
> The Data Set was downloaded from Kaggle, from the following [link](https://www.kaggle.com/datasets/lava18/google-play-store-apps/)

# 2.0 `Import Libraries:`

In [1]:
# import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# 3.0 `Exploring Data:`

In [2]:
# load data
df = pd.read_csv('../../datasets/googleplaystore.csv')

In [3]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size     Installs  Type Price Content Rating  \
0      159   19M      10,000+  Free     0       Everyone   
1      967   14M     500,000+  Free     0       Everyone   
2    87510  8.7M   5,000,000+  Free     0       Everyone   
3   215644   25M  50,000,000+  Free     0           Teen   
4      967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [4]:
df.tail()

App             Category  \
10836                               Sya9a Maroc - FR               FAMILY   
10837               Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating  Reviews                Size     Installs  Type Price  \
10836     4.5       38                 53M       5,000+  Free     0   
10837     5.0        4                3.6M         100+  Free     0   
10838     NaN        3                9.5M       1,000+  Free     0   
10839     4.5      114  Varies with device       1,000+  Free     0   
10840     4.5   398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres      Last Updated         Current Ver  \
10836       Everyone          Education     July 25, 2017                1.48   
10837       Everyone          Education      July 6, 2018                 1.0   
10838       Everyone            Medical  January 20, 2017                 1.0   
10839     Mature 17+  Books & Reference  January 19, 2015  Varies with device   
10840       Everyone          Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
10836          4.1 and up  
10837          4.1 and up  
10838          2.2 and up  
10839  Varies with device  
10840  Varies with device

In [5]:
df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10841 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10839 non-null  object 
dtypes: float64(1), int64(1), object(11)
memory usage: 1.1+ MB


In [7]:
df.describe()

Rating       Reviews
count  9367.000000  1.084100e+04
mean      4.191513  4.441119e+05
std       0.515735  2.927629e+06
min       1.000000  0.000000e+00
25%       4.000000  3.800000e+01
50%       4.300000  2.094000e+03
75%       4.500000  5.476800e+04
max       5.000000  7.815831e+07

## 3.11 `Observations:`

In [8]:
# showing all rows
pd.set_option('display.max_rows', None)

In [9]:
# showing unique values in each column with df.head()
df.columns.to_series().apply(lambda x: df[x].nunique())

App               9660
Category            33
Rating              39
Reviews           6001
Size               461
Installs            21
Type                 2
Price               92
Content Rating       6
Genres             119
Last Updated      1377
Current Ver       2831
Android Ver         33
dtype: int64

` Following are some anomalies in the data`
1. Rating has a maximum value of 19
2. Size has 'Varies with device',and  it should be numeric
3. Installs has '+' and ',', it should be numeric
4. Type has 'Free' and 'Paid'
5. Price has '$' and 'Everyone', it should be numeric
6. Content Rating has 'Adults only 18+' and 'Unrated'
7. Genres has ';'

# 4.0 `Exploring and Dealing with Columns:`

## 4.1 `App`:
### 4.1.1 `Exploring:`

In [10]:
df.App.isnull().sum()

0

In [11]:
df.App.nunique()

9660

In [12]:
# check for duplicates
df.duplicated().sum()

483

In [13]:
# DataType
df['App'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 10841 entries, 0 to 10840
Series name: App
Non-Null Count  Dtype 
--------------  ----- 
10841 non-null  object
dtypes: object(1)
memory usage: 84.8+ KB


In [14]:
df['App'].value_counts().sum()

10841

In [15]:
df[df.duplicated(keep=False, subset='App')].head()

App             Category  \
1                               Coloring book moana       ART_AND_DESIGN   
23                           Mcqueen Coloring pages       ART_AND_DESIGN   
36   UNICORN - Color By Number & Pixel Art Coloring       ART_AND_DESIGN   
42                       Textgram - write on photos       ART_AND_DESIGN   
139                            Wattpad 📖 Free Books  BOOKS_AND_REFERENCE   

     Rating  Reviews                Size      Installs  Type Price  \
1       3.9      967                 14M      500,000+  Free     0   
23      NaN       61                7.0M      100,000+  Free     0   
36      4.7     8145                 24M      500,000+  Free     0   
42      4.4   295221  Varies with device   10,000,000+  Free     0   
139     4.6  2914724  Varies with device  100,000,000+  Free     0   

    Content Rating                           Genres      Last Updated  \
1         Everyone        Art & Design;Pretend Play  January 15, 2018   
23        Everyone  Art & Design;Action & Adventure     March 7, 2018   
36        Everyone          Art & Design;Creativity    August 2, 2018   
42        Everyone                     Art & Design     July 30, 2018   
139           Teen                Books & Reference    August 1, 2018   

            Current Ver         Android Ver  
1                 2.0.0        4.0.3 and up  
23                1.0.0          4.1 and up  
36                1.0.9          4.4 and up  
42   Varies with device  Varies with device  
139  Varies with device  Varies with device

In [16]:
df[df.duplicated(keep=False, subset='App')].tail()

App       Category  Rating  Reviews  \
10715                  FarmersOnly Dating         DATING     3.0     1145   
10720  Firefox Focus: The privacy browser  COMMUNICATION     4.4    36981   
10730                         FP Notebook        MEDICAL     4.5      410   
10753      Slickdeals: Coupons & Shopping       SHOPPING     4.5    33599   
10768                                AAFP        MEDICAL     3.8       63   

       Size    Installs  Type Price Content Rating         Genres  \
10715  1.4M    100,000+  Free     0     Mature 17+         Dating   
10720  4.0M  1,000,000+  Free     0       Everyone  Communication   
10730   60M     50,000+  Free     0       Everyone        Medical   
10753   12M  1,000,000+  Free     0       Everyone       Shopping   
10768   24M     10,000+  Free     0       Everyone        Medical   

            Last Updated Current Ver Android Ver  
10715  February 25, 2016         2.2  4.0 and up  
10720       July 6, 2018         5.2  5.0 and up  
10730     March 24, 2018   2.1.0.372  4.4 and up  
10753      July 30, 2018         3.9  4.4 and up  
10768      June 22, 2018       2.3.1  5.0 and up

### 4.1.2 `Observations:`

My observations about app column are:
 1. There are 483 duplicates which are incorrect , it means that there are no duplicates in the App Column:
 2. There is no missing value
 2. There are 9660 unique values
 3. The data type is object


## 4.2 `Category`:
### 4.2.1 `Exploring:`

In [17]:
df.head(1)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   

   Reviews Size Installs  Type Price Content Rating        Genres  \
0      159  19M  10,000+  Free     0       Everyone  Art & Design   

      Last Updated Current Ver   Android Ver  
0  January 7, 2018       1.0.0  4.0.3 and up

In [18]:
df['Category'].value_counts()

Category
FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         392
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     231
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
AUTO_AND_VEHICLES        85
LIBRARIES_AND_DEMO       85
WEATHER                  82
ART_AND_DESIGN           65
EVENTS                   64
PARENTING                60
COMICS                   60
BEAUTY                   53
Name: count, dtype: int64

In [19]:
df.Category.nunique()

33

In [20]:
# Missing Values
# Dealing With Missing Values Is Compulsory
df['Category'].isnull().sum()

1

In [21]:
df.Category.duplicated().sum()

10807

In [22]:
df[df.duplicated(subset='Category', keep=False)].head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size     Installs  Type Price Content Rating  \
0      159   19M      10,000+  Free     0       Everyone   
1      967   14M     500,000+  Free     0       Everyone   
2    87510  8.7M   5,000,000+  Free     0       Everyone   
3   215644   25M  50,000,000+  Free     0           Teen   
4      967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [23]:
df[df.duplicated(subset='Category', keep=False)].tail()

App             Category  \
10836                               Sya9a Maroc - FR               FAMILY   
10837               Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating  Reviews                Size     Installs  Type Price  \
10836     4.5       38                 53M       5,000+  Free     0   
10837     5.0        4                3.6M         100+  Free     0   
10838     NaN        3                9.5M       1,000+  Free     0   
10839     4.5      114  Varies with device       1,000+  Free     0   
10840     4.5   398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres      Last Updated         Current Ver  \
10836       Everyone          Education     July 25, 2017                1.48   
10837       Everyone          Education      July 6, 2018                 1.0   
10838       Everyone            Medical  January 20, 2017                 1.0   
10839     Mature 17+  Books & Reference  January 19, 2015  Varies with device   
10840       Everyone          Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
10836          4.1 and up  
10837          4.1 and up  
10838          2.2 and up  
10839  Varies with device  
10840  Varies with device

### 4.2.2 `Observations:`

Following are my Observations about the Category Column:
1. Total categories in this dataset are 33
2. Total number of missing value is 1
3. There are no Duplicate rows, the duplicates it shows are not correct as many apps can have the same category, and there are no duplicate app in the dataset
4. Top Categories based on number of Apps are the Following:
- FAMILY     = 1972
- GAME        = 1144
- TOOLS        = 843
- MEDICAL      = 463
- BUSINESS     = 460

## 4.3 `Rating`:
### 4.3.1 `Exploring:`

In [ ]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews Size  Installs  Type Price Content Rating  \
0      159  19M   10,000+  Free     0       Everyone   
1      967  14M  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

In [ ]:
df.Rating.isnull().sum()

1474

In [32]:
df['Rating'].nunique()

39

In [39]:
df['Rating'].value_counts()

Rating
4.4    1109
4.3    1076
4.5    1038
4.2     952
4.6     823
4.1     708
4.0     568
4.7     499
3.9     386
3.8     303
5.0     274
3.7     239
4.8     234
3.6     174
3.5     163
3.4     128
3.3     102
4.9      87
3.0      83
3.1      69
3.2      64
2.9      45
2.8      42
2.7      25
2.6      25
2.5      21
2.3      20
2.4      19
1.0      16
1.9      14
2.2      14
2.0      12
1.7       8
1.8       8
2.1       8
1.6       4
1.4       3
1.5       3
1.2       1
Name: count, dtype: int64

In [47]:
# number of apps with rating 4 and above
df[df['Rating'] >= 4].shape[0]

7368

In [51]:
df[(df['Rating'] < 4) & (df['Rating'] > 2)].shape[0]


1930

In [48]:
# number of apps with rating 2 and below
df[df['Rating']<=2].shape[0]

69

### 4.3.2 `Observations:`

I have done the following observations from the Rating Column:
1. Rating have 1474 Missing Values
2. Rating Column have 39 unique Values
3. Apps with rating 4 and above are `7368`
4. Apps with rating 2 to 4 are `1930`
5. Apps with rating 2 and below are `69`

## 4.4 `Reviews`:
### 4.4.1 `Exploring:`      

In [52]:
df.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews Size  Installs  Type Price Content Rating  \
0      159  19M   10,000+  Free     0       Everyone   
1      967  14M  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

In [53]:
df['Reviews'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 10841 entries, 0 to 10840
Series name: Reviews
Non-Null Count  Dtype
--------------  -----
10841 non-null  int64
dtypes: int64(1)
memory usage: 84.8 KB


In [55]:
df['Reviews'].isnull().sum()

0

In [56]:
df['Reviews'].nunique()

6001

In [59]:
df[df['Reviews']>10000].shape[0]

4280

### 4.4.2 `Observations:`

## 4.5 `Size`:
### 4.5.1 `Exploring:`

### 4.5.2 `Observations:`

## 4.6 `Installs`:
### 4.6.1 `Exploring:`

### 4.6.2 `Observations:`

## 4.7 `Type`:
### 4.7.1 `Exploring:`

### 4.7.2 `Observations:`

## 4.8 `Price`:
### 4.8.1 `Exploring:`

### 4.8.2 `Observations:`

## 4.9 `Content Rating`:
### 4.9.1 `Exploring:`

### 4.9.2 `Observations:`

## 4.10 `Genres`:
### 4.10.1 `Exploring:`

### 4.10.2 `Observations:`

## 4.11 `Last Updated`:
### 4.11.1 `Exploring:`

### 4.11.2 `Observations:`

## 4.12 `Current Version`:
### 4.12.1 `Exploring:`

### 4.12.2 `Observations:`

## 4.13 `Andriod Version`:
### 4.13.1 `Exploring:`

### 4.13.2 `Observations:`